### ISSUE TITLE
rethinking if-else

### ISSUE THREAD

**Issue by gnudles** [dart-lang/language#1723]
The case when I have multiple time consuming function returning bool, and I need to execute code special code for each one and also code if any of them is true.
bool f1();// time consuming
bool f2();// also
 making cache of the results adds unnecessary lines
of code.
current status:
if (f1())
{
  handle f1...
}
if ( f2())
{
  handle f2...
}
if ( f1() || f2())
{
  handle common...
}
my suggestion is to allow the following syntax:
keywords chosen by me are tentative.
either (f1())
{
handle f1..
}
or (f2())
{
 handle f2...
}
     ///or ( f3()){}
do {
 handle common
}
...
maybe think also on the && case where need to execute code only if all are true...

**Comment by Levi-Lesches** [comment]
How about saving the results to a local variable?

```dart
Future<bool> getNetworkReply(String message) { } 

Future<void> doSomething() async {
  final bool isSignedIn = await getNetworkReply("sign-in");
  final bool isAdmin = await getNetworkReply("is-admin");

  if (isSignedIn) 
    removeSignInScreen();
  if (isAdmin) 
    prepareAdminAPI();
  if (isSignedIn && isAdmin)
    showAdminPage();
}
```

**Comment by gnudles** [comment]
This is the obvious, and I mentioned it "making cache of the results adds unnecessary lines
of code.".
I just think that programming languages should evolve to be more readable, and consume less lines of code.
this case is very common and every time I encounter it I feel bad about my code becoming ugly. As a philosophy, I think that a boolean value should be checked only once.

**Comment by gnudles** [comment]
if(b1)
{
}
or(b2)
{
}
if all
{
}
if any
{
}

**Comment by lrhn** [comment]
That syntax seems special tailored for two values. What if you have three? Four? Doesn't seem like something which generalizes.

I'd probably go with a helper implementation like:
```dart
swtich (combine(f1(), f2())) {
  case Combined.first:
    ...
  case Combined.second:
    ...
  case Combined.both:
    ...
  case Combined.neither:
    ...
}
/// with
enum Combined {
   neither, first, second, both;
}
Combined combine(bool b1, bool b2) => Combined.values[(b1 ? 1 : 0) | (b2 ? 2 : 0)];
```

That is, if you really need to reuse something like that so often it's worth considering a language feature for it, then that reusable implementation seems reasonable too.

**Comment by gnudles** [comment]
First, it can be generalized for more than two values. it will just do && or || between all of them.
Second, I liked your implementation.
I think this is a place the if-else could evolve into, to make code appear more tidy. I just had it on my mind and thought it would be nice to have

**Comment by munificent** [comment]
My experience is that users already struggle to write and read complex boolean conditional logic and control flow correctly even with the handful of simple, familiar constructs we already have. Adding more constructs *might* let them write simpler code, but my suspicion is that the cost to learn those new constructs would outweigh the benefits of having them.

I think in many cases it's better to hoist the repeated computation into a local variable. Yes, the code is a little verbose, but it's more verbose in a valuable way: you get to assign a meaningful *name* that explains what these computations represents.
